File: sync-bridge-rust/src/handlers/health.rs
// src/handlers/health.rs
use axum::{extract::State, http::StatusCode, response::IntoResponse};
use chrono::Utc;
use serde_json::json;
use sqlx::SqlitePool;

pub async fn health_handler(State(pool): State<SqlitePool>) -> impl IntoResponse {
    let mut read_ok = false;
    let mut write_ok = false;

    // 1. Perform database read check
    if sqlx::query("SELECT 1").execute(&pool).await.is_ok() {
        read_ok = true;
    }

    // 2. Perform database write check: insert a temp customer and delete it
    let email = format!("healthcheck-{}@example.com", uuid::Uuid::new_v4());
    let write_res = sqlx::query(
        "INSERT INTO customers (email, first_name, last_name, default_currency) VALUES (?, 'Health', 'Check', 'USD')"
    )
    .bind(&email)
    .execute(&pool)
    .await;

    if write_res.is_ok() {
        // Clean up immediately
        let _ = sqlx::query("DELETE FROM customers WHERE email = ?")
            .bind(&email)
            .execute(&pool)
            .await;
        write_ok = true;
    }

    let is_healthy = read_ok && write_ok;
    let status = if is_healthy {
        StatusCode::OK
    } else {
        StatusCode::SERVICE_UNAVAILABLE
    };

    let message = if is_healthy {
        "Service is healthy"
    } else {
        "Service is unhealthy"
    };

    let data = json!({
        "read": read_ok,
        "write": write_ok,
        "timestamp": Utc::now().to_rfc3339()
    });

    let body = json!({
        "status": status.as_u16(),
        "message": message,
        "data": data
    });

    (status, axum::Json(body))
}

File: sync-bridge-go/internal/handlers/health.go
package handlers

import (
	"database/sql"
	"fmt"
	"net/http"
	"time"

	"github.com/gin-gonic/gin"
	"github.com/google/uuid"
)

// HealthResponse represents the health check response payload
type HealthResponse struct {
	Status  int             `json:"status" example:"200"`
	Message string          `json:"message" example:"Service is healthy"`
	Data    HealthCheckData `json:"data"`
}

// HealthCheckData represents the detail status of system components
type HealthCheckData struct {
	Read      bool   `json:"read" example:"true"`
	Write     bool   `json:"write" example:"true"`
	Timestamp string `json:"timestamp" example:"2026-06-03T07:46:58Z"`
}

// HealthHandler godoc
// @Summary      System health check
// @Description  Verifies basic server operation and database read/write capabilities
// @Tags         system
// @Produce      json
// @Success      200  {object}  HealthResponse
// @Failure      503  {object}  HealthResponse
// @Router       /api/v1/healthz [get]
func HealthHandler(db *sql.DB) gin.HandlerFunc {
	return func(c *gin.Context) {
		readOk := false
		writeOk := false

		// 1. Perform database read check
		var val int
		err := db.QueryRow("SELECT 1").Scan(&val)
		if err == nil {
			readOk = true
		}

		// 2. Perform database write check: insert a temp customer and delete it
		email := fmt.Sprintf("healthcheck-%s@example.com", uuid.New().String())
		_, writeErr := db.Exec(
			"INSERT INTO customers (email, first_name, last_name, default_currency) VALUES (?, 'Health', 'Check', 'USD')",
			email,
		)

		if writeErr == nil {
			// Clean up immediately
			_, _ = db.Exec("DELETE FROM customers WHERE email = ?", email)
			writeOk = true
		}

		isHealthy := readOk && writeOk
		status := http.StatusOK
		message := "Service is healthy"
		if !isHealthy {
			status = http.StatusServiceUnavailable
			message = "Service is unhealthy"
		}

		c.JSON(status, gin.H{
			"status":  status,
			"message": message,
			"data": gin.H{
				"read":      readOk,
				"write":     writeOk,
				"timestamp": time.Now().UTC().Format(time.RFC3339),
			},
		})
	}
}
